lonestar/analytics/cpu/k-shortest-paths/SSSP.cpp

```diff
@@ -0,0 +1,457 @@
+  katana::FixedSizeAllocator<Path> allocator_;
+};
+
+template <typename Item, typename PushWrap, typename EdgeRange>
+bool
+CheckReachabilityAsync(
+    Graph* graph, const GNode& source, const PushWrap& pushWrap,
+    const EdgeRange& edgeRange) {
+  using FIFO = katana::PerSocketChunkFIFO<kChunkSize>;
+  using WL = FIFO;
+
+  using Loop = typename std::conditional<
+      true, katana::ForEach, katana::WhileQ<katana::SerFIFO<Item>>>::type;
+
+  Loop loop;
+
+  graph->GetData<NodeCount>(source) = 1;
+  katana::InsertBag<Item> initBag;
+
+  pushWrap(initBag, source, 1, "parallel");
+
+  loop(
+      katana::iterate(initBag),
+      [&](const Item& item, auto& ctx) {
+        for (auto ii : edgeRange(item)) {
+          GNode dst = *(graph->GetEdgeDest(ii));
+          if (graph->GetData<NodeCount>(dst) == 0) {
+            graph->GetData<NodeCount>(dst) = 1;
+            pushWrap(ctx, dst, 1);
+          }
+        }
+      },
+      katana::wl<WL>(), katana::loopname("runBFS"),
+      katana::disable_conflict_detection());
+
+  if (graph->GetData<NodeCount>(reportNode) == 0) {
+    return false;
+  }
+
+  katana::do_all(katana::iterate(*graph), [&graph](GNode n) {
+    graph->GetData<NodeCount>(n) = 0;
+  });
+
+  return true;
+}
+
+bool
+CheckReachabilitySync(Graph* graph, const GNode& source) {
+  katana::InsertBag<GNode> current_bag;
+  katana::InsertBag<GNode> next_bag;
+
+  current_bag.push(source);
+  graph->GetData<NodeCount>(source) = 1;
+
+  while (current_bag.begin() != current_bag.end()) {
+    katana::do_all(
+        katana::iterate(current_bag),
+        [&](GNode n) {
+          for (auto edge : graph->edges(n)) {
+            auto dest = *(graph->GetEdgeDest(edge));
+            if (graph->GetData<NodeCount>(dest) == 0) {
+              graph->GetData<NodeCount>(dest) = 1;
+              next_bag.push(dest);
+            }
+          }
+        },
+        katana::steal());
+
+    current_bag.clear();
+    std::swap(current_bag, next_bag);
+  }
+
+  if (graph->GetData<NodeCount>(reportNode) == 0) {
+    return false;
+  }
+
+  katana::do_all(katana::iterate(*graph), [&graph](GNode n) {
+    graph->GetData<NodeCount>(n) = 0;
+  });
+
+  return true;
+}
+
+//delta stepping implementation for finding a shortest path from source to report node
+template <typename Item, typename OBIMTy, typename PushWrap, typename EdgeRange>
+void
+DeltaStepAlgo(
+    Graph* graph, const GNode& source, const PushWrap& pushWrap,
+    const EdgeRange& edgeRange,
+    katana::InsertBag<std::pair<uint32_t, Path*>>* report_paths_bag,
+    katana::InsertBag<Path*>* path_pointers, PathAlloc& path_alloc) {
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> bad_work;
+  //! [reducible for self-defined stats]
+  katana::GAccumulator<size_t> wl_empty_work;
+
+  graph->GetData<NodeCount>(source) = 1;
+
+  katana::InsertBag<Item> init_bag;
+
+  //Path* path = new Path();
+  Path* path;
+  path_alloc.NewPath();
+  path->last = nullptr;
```

Comment:
Drop line. It is not required with the default initialization I mentioned above.

However, what is the value of parent for this path?

Reply:
this path pointer corresponds to the source node, hence no parent to keep track of.

Reply:
It's not good to leave the parent uninitialized. Maybe just assign it to the source?

Reply:
ok, I will do that